### ISSUE TITLE
SqlServer-2.1 - InvalidCastException (Int32 to Boolean)

### ISSUE THREAD

**Issue by huancz** [aspnet/EntityFrameworkCore#12280]
Simple .Select projection throws in Microsoft.EntityFrameworkCore.SqlServer 2.1 (worked fine in 2.0 or Memory DB of any version)

```
System.InvalidOperationException: An exception occurred while reading a database value. The expected type was 'System.Boolean' but the actual value was of type 'System.Int32'. ---> System.InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Boolean'.
         at System.Data.SqlClient.SqlBuffer.get_Boolean()
         at lambda_method(Closure , DbDataReader )
         --- End of inner exception stack trace ---
         at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.ThrowReadValueException[TValue](Exception excepti
on, Object value, IPropertyBase property)
         at lambda_method(Closure , DbDataReader )
         at Microsoft.EntityFrameworkCore.Storage.Internal.TypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
         at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
         at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptorUnhandled Exception:.MoveNext()
```

### Steps to reproduce

Full repro as attachment, core of the issue is this snippet:
```c#
List<string> selectedNames = new List<string>();
var selectedList = context.Set<Entity>()
    .Select(x => new
    {
        Name = x.Name,
        Selected = selectedNames != null && selectedNames.Contains(x.Name)
    })
    .ToList()
```
It is specific to .Contains on empty list. Various workarounds:
- selectedNames = null
- selectedNames.Add("nonsense")
- rewriting the projection to read "Selected = selectedNames != null && selectedNames.Any() && selectedNames.Contains(x.Name)"

This seems like repeat of issue #2246, but that one was fixed two years ago, now it's back again in 2.1. I couldn't find anything similar in currently open issues.

### Further technical details
EF Core version: 2.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.7.3

[repro-invalid-cast.zip](https://github.com/aspnet/EntityFrameworkCore/files/2080909/repro-invalid-cast.zip)

**Comment by ajcvickers** [comment]
Possible duplicate of #12175 since they both involve empty Contains. /cc @maumar @smitpatel

**Comment by smitpatel** [comment]
Duplicate of #12175

**Comment by smitpatel** [comment]
Verified fixed in nightlies

**Comment by ajcvickers** [comment]
This is approved for 2.1.3. Do not merge yet; branch is expected to open Monday.

Given these issues have been parked for some time, please be careful to ensure the correct commits get into the correct branches.

**Comment by ajcvickers** [comment]
@ajcvickers This issue is approved for patch and the release\2.1 branch is now open for merging. Please ensure:
* The appropriate changes get into both the release and dev branches
* Quirking is included for the release branch only

**Comment by maumar** [comment]
Merged to 2.1 in 77512c451480d52f9b4e2970a8ba72d9f38c0141
